storage-proofs/src/piece_inclusion_proof.rs

```diff
@@ -1,255 +1,473 @@
+#[allow(dead_code)]
+/// Generate comm_p from bytes
+fn generate_comm_p<H: Hasher>(data: &[u8]) -> Result<H::Domain> {
+    let mut domain_data = Vec::new();
+    for d in data.chunks(32) {
+        domain_data.push(<H::Domain as Domain>::try_from_bytes(d)?)
+    }
+
+    Ok(compute_comm_p::<H>(&domain_data))
 }
 
-/// verify_file_inclusion_proofs returns true iff each provided piece is proved with respect to root
-/// by the corresponding (by index) proof.
-pub fn verify_file_inclusion_proofs<H: Hasher>(
-    root: &H::Domain,
-    proofs: &[PieceInclusionProof<H>],
-    pieces: &[&[u8]],
-) -> bool {
-    proofs
+#[allow(dead_code)]
+/// Generate comm_p from bytes and return it as bytes.
+fn generate_comm_p_bytes<H: Hasher>(data: &[u8]) -> Result<Fr32Vec> {
+    let comm_p = generate_comm_p::<H>(data)?;
+
+    Ok(H::Domain::into_bytes(&comm_p))
+}
+
+pub fn piece_inclusion_proofs<H: Hasher>(
+    piece_specs: &[PieceSpec],
+    tree: &MerkleTree<H::Domain, H::Function>,
+) -> Result<Vec<PieceInclusionProof<H>>> {
+    piece_specs
         .iter()
-        .zip(pieces)
-        .all(|(proof, piece)| proof.verify(root, piece))
+        .map(|piece_spec| PieceInclusionProof::new((*piece_spec).clone(), tree))
+        .collect()
+}
+
+impl<H: Hasher> PieceInclusionProof<H> {
+    pub fn new(
+        piece_spec: PieceSpec,
+        tree: &MerkleTree<H::Domain, H::Function>,
+    ) -> Result<PieceInclusionProof<H>> {
+        let PieceSpec {
+            comm_p,
+            position: first_leaf,
+            length: leaf_count,
+        } = piece_spec;
+
+        let last_leaf = first_leaf + (leaf_count - 1);
+
+        // For now, we only handled aligned pieces.
+        if !piece_spec.is_aligned(tree.len()) {
+            return Err(Error::UnalignedPiece);
+        }
+
+        let first_proof = tree.gen_proof(first_leaf);
+        let last_proof = tree.gen_proof(last_leaf);
+
+        // Including first leaf (item).
+        let proof_length = piece_spec.height() + 1;
+
+        // Find the first common hash (at same position) in first and last proof,
+        // then discard all previous hashes and path bits, and create a new proof containing
+        // only the common part.
+        //
+        // For an aligned piece, there is guaranteed to be a common element.
+        // That is, the merkle root of the piece must be a node of the tree (and therefore in the paths
+        // of all leaves of the piece (including the first and last).
+        for (i, (a, b)) in first_proof
+            .lemma()
+            .iter()
+            .zip(last_proof.lemma().iter())
+            .enumerate()
+        {
+            if (a == b) && (i == proof_length) {
+                let proof_elements = if i > first_proof.path().len() {
+                    vec![]
+                } else {
+                    let mut lemma = first_proof.lemma()[i..].to_vec();
+                    lemma.insert(0, H::Domain::try_from_bytes(&comm_p)?);
```

Comment:
this could be a bit more efficient

```rust
let mut lemma = Vec::with_capacity(first_proof.lemma().len() - i + 1);
lemma.push(H::Domain::try_from_bytes(&comm_p)?);
lemma.extend_from_slice(&first_proof.lemma()[i..]);
```

Reply:
Actually, the whole song and dance can be simplified out of existence as the last stage of refactoring. Fix incoming.